// @loopback/context/dist8/src/index.js
"use strict";
// Copyright IBM Corp. 2017,2018. All Rights Reserved.
// Node module: @loopback/context
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("@loopback/metadata"));
var value_promise_1 = require("./value-promise");
exports.isPromiseLike = value_promise_1.isPromiseLike;
exports.resolveList = value_promise_1.resolveList;
exports.resolveMap = value_promise_1.resolveMap;
exports.resolveUntil = value_promise_1.resolveUntil;
exports.transformValueOrPromise = value_promise_1.transformValueOrPromise;
exports.tryWithFinally = value_promise_1.tryWithFinally;
exports.getDeepProperty = value_promise_1.getDeepProperty;
var binding_1 = require("./binding");
exports.Binding = binding_1.Binding;
exports.BindingScope = binding_1.BindingScope;
exports.BindingType = binding_1.BindingType;
var context_1 = require("./context");
exports.Context = context_1.Context;
var binding_key_1 = require("./binding-key");
exports.BindingKey = binding_key_1.BindingKey;
var resolution_session_1 = require("./resolution-session");
exports.ResolutionSession = resolution_session_1.ResolutionSession;
var inject_1 = require("./inject");
exports.inject = inject_1.inject;
var resolver_1 = require("./resolver");
exports.instantiateClass = resolver_1.instantiateClass;
exports.invokeMethod = resolver_1.invokeMethod;
// internals for testing
var inject_2 = require("./inject");
exports.describeInjectedArguments = inject_2.describeInjectedArguments;
exports.describeInjectedProperties = inject_2.describeInjectedProperties;
