// @loopback/rest/dist10/src/index.js
"use strict";
// Copyright IBM Corp. 2017,2018. All Rights Reserved.
// Node module: @loopback/rest
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var routing_table_1 = require("./router/routing-table");
exports.RoutingTable = routing_table_1.RoutingTable;
exports.Route = routing_table_1.Route;
exports.ControllerRoute = routing_table_1.ControllerRoute;
exports.createResolvedRoute = routing_table_1.createResolvedRoute;
exports.createControllerFactoryForBinding = routing_table_1.createControllerFactoryForBinding;
exports.createControllerFactoryForClass = routing_table_1.createControllerFactoryForClass;
exports.createControllerFactoryForInstance = routing_table_1.createControllerFactoryForInstance;
__export(require("./providers"));
__export(require("./parser"));
__export(require("./writer"));
__export(require("./http-handler"));
__export(require("./request-context"));
__export(require("./keys"));
__export(require("./rest.application"));
__export(require("./rest.component"));
__export(require("./rest.server"));
__export(require("./sequence"));
__export(require("./rest-http-error"));
// export all errors from external http-errors package
const HttpErrors = require("http-errors");
exports.HttpErrors = HttpErrors;
__export(require("@loopback/openapi-v3"));
