// @loopback/repository/dist8/src/repositories/relation.repository.js
"use strict";
// Copyright IBM Corp. 2017,2018. All Rights Reserved.
// Node module: @loopback/example-todo
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
const constraint_utils_1 = require("./constraint-utils");
class DefaultHasManyEntityCrudRepository {
    /**
     * Constructor of DefaultHasManyEntityCrudRepository
     * @param targetRepository the related target model repository instance
     * @param constraint the key value pair representing foreign key name to constrain
     * the target repository instance
     */
    constructor(targetRepository, constraint) {
        this.targetRepository = targetRepository;
        this.constraint = constraint;
    }
    async create(targetModelData, options) {
        return await this.targetRepository.create(constraint_utils_1.constrainDataObject(targetModelData, this.constraint), options);
    }
    async find(filter, options) {
        return await this.targetRepository.find(constraint_utils_1.constrainFilter(filter, this.constraint), options);
    }
    async delete(where, options) {
        return await this.targetRepository.deleteAll(constraint_utils_1.constrainWhere(where, this.constraint), options);
    }
    async patch(dataObject, where, options) {
        return await this.targetRepository.updateAll(constraint_utils_1.constrainDataObject(dataObject, this.constraint), constraint_utils_1.constrainWhere(where, this.constraint), options);
    }
}
exports.DefaultHasManyEntityCrudRepository = DefaultHasManyEntityCrudRepository;
